src/Integration/LocalServices/SolutionBindingInformationProvider.cs

```diff
@@ -0,0 +1,126 @@
+﻿//-----------------------------------------------------------------------
+// <copyright file="SolutionBindingInformationProvider.cs" company="SonarSource SA and Microsoft Corporation">
+//   Copyright (c) SonarSource SA and Microsoft Corporation.  All rights reserved.
+//   Licensed under the MIT License. See License.txt in the project root for license information.
+// </copyright>
+//-----------------------------------------------------------------------
+
+using EnvDTE;
+using Microsoft.VisualStudio.CodeAnalysis.RuleSets;
+using SonarLint.VisualStudio.Integration.Binding;
+using SonarLint.VisualStudio.Integration.Persistence;
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+
+namespace SonarLint.VisualStudio.Integration
+{
+    internal class SolutionBindingInformationProvider : ISolutionBindingInformationProvider
+    {
+        private readonly IServiceProvider serviceProvider;
+
+        public SolutionBindingInformationProvider(IServiceProvider serviceProvider)
+        {
+            if (serviceProvider == null)
+            {
+                throw new ArgumentNullException(nameof(serviceProvider));
+            }
+
+            this.serviceProvider = serviceProvider;
+        }
+
+        #region ISolutionBindingInformationProvider
+        public IEnumerable<Project> GetBoundProjects()
+        {
+            var bindingSerializer = this.serviceProvider.GetService<ISolutionBindingSerializer>();
+            bindingSerializer.AssertLocalServiceIsNotNull();
+
+            // We only have bound projects if the solution has persisted solution binding
+            BoundSonarQubeProject binding = bindingSerializer.ReadSolutionBinding();
+            if (binding == null)
+            {
+                return Enumerable.Empty<Project>();
+            }
+
+            var projectSystem = this.serviceProvider.GetService<IProjectSystemHelper>();
+            projectSystem.AssertLocalServiceIsNotNull();
+
+            return projectSystem.GetFilteredSolutionProjects()
+                .Where(p => this.IsFullyBoundProject(binding, p));
+        }
+
+        public IEnumerable<Project> GetUnboundProjects()
+        {
+            var projectSystem = this.serviceProvider.GetService<IProjectSystemHelper>();
+            projectSystem.AssertLocalServiceIsNotNull();
+
+            return projectSystem.GetFilteredSolutionProjects().Except(this.GetBoundProjects());
+        }
+        #endregion
+
+        #region Non-public API
+        private bool IsFullyBoundProject(BoundSonarQubeProject binding, Project project)
+        {
+            Debug.Assert(binding != null);
+            Debug.Assert(project != null);
+
+            // If solution is not bound/has a missing ruleset, no need to go further
+            RuleSet sonarQubeRuleSet = this.FindSonarQubeSolutionRuleSet(binding, project);
```

Comment:
So we're going to check for the solution rule set every time per project? Seems a little wasteful.
Yes, the point here is that there _might_ be two different rule sets (one for C# and one for VB.NET) but 1) we'll still be checking potentially multiple times if we have >1 project of the same language, and 2) we actually only support C# at the moment so in practice, yes we will be doing more work.
Probably not worth changing this check now though.